src/client.js

```diff
@@ -0,0 +1,157 @@
+const EventEmitter = require('events');
+const Long = require('long');
+const Parser = require('./parser');
+const decrypt = require('./utils/decrypt');
+const path = require('path');
+const tls = require('tls');
+const {checkIn} = require('./gcm');
+const {kMCSVersion, kLoginRequestTag} = require('./constants');
+const {load} = require('protobufjs');
+
+const HOST = 'mtalk.google.com';
+const PORT = 5228;
+const MAX_RETRY_TIMEOUT = 15;
+
+let proto = null;
+
+module.exports = class Client extends EventEmitter {
+  static async init () {
+    if (proto) {
+      return;
+    }
+    proto = await load(path.resolve(__dirname, 'mcs.proto'));
+  }
+
+  constructor(credentials, persistentIds) {
+    super();
+    this._credentials = credentials;
+    this._persistentIds = persistentIds || [];
+    this._retryCount = 0;
+    this._onSocketConnect = this._onSocketConnect.bind(this);
+    this._onSocketClose = this._onSocketClose.bind(this);
+    this._onSocketError = this._onSocketError.bind(this);
+    this._onDataMessage = this._onDataMessage.bind(this);
+    this._onParserError = this._onParserError.bind(this);
+  }
+
+  async connect() {
+    await Client.init();
+    await this._checkIn();
+    this._connect();
+    // can happen if the socket immediately closes after being created
+    if (!this._socket) { return }
+    await Parser.init();
+    // can happen if the socket immediately closes after being created
+    if (!this._socket) { return }
+    this._parser = new Parser(this._socket);
+    this._parser.on('dataMessage', this._onDataMessage);
+    this._parser.on('error', this._onParserError);
+  }
+
+  destroy() {
+    this._destroy();
+  }
+
+  async _checkIn() {
+    return checkIn(this._credentials.gcm.androidId, this._credentials.gcm.securityToken);
+  }
+
+  _connect() {
+    this._socket = new tls.TLSSocket();
+    this._socket.setKeepAlive(true);
+    this._socket.on('connect', this._onSocketConnect);
+    this._socket.on('close', this._onSocketClose);
+    this._socket.on('error', this._onSocketError);
+    this._socket.connect({host : HOST, port : PORT});
+    this._socket.write(this._loginBuffer());
+  }
+
+  _destroy() {
+    clearTimeout(this._retryTimeout);
+    if (this._socket) {
+      this._socket.removeListener('connect', this._onSocketConnect);
+      this._socket.removeListener('close', this._onSocketClose);
+      this._socket.removeListener('error', this._onSocketError);
+      this._socket.destroy();
+      this._socket = null;
+    }
+    if (this._parser) {
+      this._parser.removeListener('dataMessage', this._onDataMessage);
```

Comment:
You're also listening to `error` from line 48.

Reply:
Actually, there's still this one.

Reply:
ah good catch! wil fix